pkg/ipam-controller/controllers/cidrpool/cidrpool.go

```diff
@@ -0,0 +1,289 @@
+// create new allocation for the node, returns nil if can't generate a subnet because lack of free prefixes
+func (r *CIDRPoolReconciler) getNewAllocationForNode(subnetGen func() *net.IPNet, node string,
+	pool *ipamv1alpha1.CIDRPool,
+	allocByPrefix map[string]*ipamv1alpha1.CIDRPoolAllocation,
+	staticAllocByPrefix map[string]*ipamv1alpha1.CIDRPoolStaticAllocation) *ipamv1alpha1.CIDRPoolAllocation {
+	var nodePrefix *net.IPNet
+	for {
+		nodePrefix = subnetGen()
+		if nodePrefix == nil {
+			break
+		}
+		if _, alreadyAllocated := allocByPrefix[nodePrefix.String()]; alreadyAllocated {
+			// subnet is already allocated, try next subnet
+			continue
+		}
+		if _, alreadyAllocated := staticAllocByPrefix[nodePrefix.String()]; alreadyAllocated {
+			// static allocation for the subnet exist (not for the current node), try next subnet
+			continue
+		}
+		break
+	}
+	if nodePrefix == nil {
+		return nil
+	}
+	return &ipamv1alpha1.CIDRPoolAllocation{
+		NodeName: node,
+		Prefix:   nodePrefix.String(),
+		Gateway:  ipamv1alpha1.GetGatewayForSubnet(nodePrefix, pool.Spec.GatewayIndex),
+	}
+}
+
+// builds CIDRPoolAllocation from the CIDRPoolStaticAllocation if it exist for the node
+func (r *CIDRPoolReconciler) getStaticAllocationForNodeIfExist(node string, pool *ipamv1alpha1.CIDRPool,
+	staticAllocationsByNodeName map[string]*ipamv1alpha1.CIDRPoolStaticAllocation) *ipamv1alpha1.CIDRPoolAllocation {
+	staticAlloc, staticAllocationFound := staticAllocationsByNodeName[node]
+	if !staticAllocationFound {
+		return nil
+	}
+	gateway := staticAlloc.Gateway
+	if gateway == "" {
+		// prefix is already validated
+		_, nodeSubnet, _ := net.ParseCIDR(staticAlloc.Prefix)
+		gateway = ipamv1alpha1.GetGatewayForSubnet(nodeSubnet, pool.Spec.GatewayIndex)
+	}
+	return &ipamv1alpha1.CIDRPoolAllocation{
+		NodeName: staticAlloc.NodeName,
+		Prefix:   staticAlloc.Prefix,
+		Gateway:  gateway,
+	}
+}
+
+// returns two maps with valid existing allocations,
+// first map contains nodeName to allocations mapping,
+// second map contains prefix to allocation mapping
+func (r *CIDRPoolReconciler) getValidExistingAllocationsMaps(ctx context.Context,
+	pool *ipamv1alpha1.CIDRPool, nodes []string) (
+	map[string]*ipamv1alpha1.CIDRPoolAllocation, map[string]*ipamv1alpha1.CIDRPoolAllocation) {
+	reqLog := log.FromContext(ctx)
+	nodeMap := make(map[string]*ipamv1alpha1.CIDRPoolAllocation, len(pool.Status.Allocations))
+	prefixMap := make(map[string]*ipamv1alpha1.CIDRPoolAllocation, len(pool.Status.Allocations))
+
+	matchingNodes := make(map[string]struct{}, len(nodes))
+	for _, n := range nodes {
+		matchingNodes[n] = struct{}{}
+	}
+
+	for i := range pool.Status.Allocations {
+		alloc := &pool.Status.Allocations[i]
+		if _, nodeExist := matchingNodes[alloc.NodeName]; !nodeExist {
+			reqLog.Info("node doesn't match the pool, discard the allocation", "node",
+				alloc.NodeName, "prefix", alloc.Prefix, "gateway", alloc.Gateway)
+			continue
+		}
+		errList := alloc.Validate(pool)
```

Comment:
Yes, we use this logic to discard allocations that are not valid anymore. We don't want to reset all allocations in case if the spec was update, instead we want to validate each allocation one by one to be able to keep allocations that are still valid.